src/thing_creature.c

```diff
@@ -7124,6 +7124,39 @@ TbBool creature_can_be_transferred(const struct Thing* thing)
     return ((get_creature_model_flags(thing) & CMF_NoTransfer) == 0);
 }
 
+/* Returns a random creature kind with model flags as argument. */
+ThingModel get_random_creature_kind_with_model_flags(unsigned long model_flags)
+{
+    // Array to store the IDs of creatures kinds with model flags.
+    ThingModel creature_kind_with_model_flags_array[CREATURE_TYPES_MAX];
+    // Counter for the number of creatures kinds found.
+    short creature_kind_with_model_flags_count = 0;
+    // Loop through all available creatures kinds.
+    for (ThingModel crkind = 0; crkind < game.conf.crtr_conf.model_count; crkind++)
+    {
+        // Check if the creature kind has the flag.
+        if (flag_is_set(game.conf.crtr_conf.model[crkind].model_flags, model_flags))
+        {
+            // Ensure we don't exceed the maximum array size.
+            if (creature_kind_with_model_flags_count < CREATURE_TYPES_MAX)
+            {
+                // Add the creature kind to the array.
+                creature_kind_with_model_flags_array[creature_kind_with_model_flags_count++] = crkind;
+            } else {
+                break;
+            }
+        }
+    }
+    if (creature_kind_with_model_flags_count > 0)
+    {
+        // Get a random creature kind from the list.
+        short random_idx = GAME_RANDOM(creature_kind_with_model_flags_count);
+        return creature_kind_with_model_flags_array[random_idx];
+    }
+    // Return -1 if no suitable creature kind is found.
+    return -1;
+}
+
```

Comment:
this seems quite inefficient, a more optimal way to do same thing would be, just start looking at a random spot in the list, and return the first one found instead of looping trough everything caching the results etc

Reply:
something like 
rand = RANDOM(models_count)
for(i = 0,models_count,i++)
  mdl_idx = (i + rand)%models_count
  
  then just return when you find one where mdl_idx is a match

Reply:
or thinking again, I guess that would favour models that have a bunch of ones that aren't a match so nvm